src/qtt/algorithms/markov_chain.py

```diff
@@ -1,74 +1,88 @@
+def _solve_least_squares(a: np.ndarray, b: np.ndarray) -> np.ndarray:
     rcond = None
     solution = np.linalg.lstsq(a, b, rcond=rcond)[0]
     return solution
 
 
+@dataclass
 class ChoiceGenerator:
-    """ Class to generate random elements with weighted selection """
+    """ Class to generate random elements with weighted selection
 
-    def __init__(self, number_of_states, cum_weights, block_size=5000):
-        """ Class to generate random elements with weighted selection
+    This is a replacement for random.choices that is efficient when a large number of choices has to be generated.
 
-        This is a replacement for random.choices that is efficient when a large number of choices has to be generated.
+    Args:
+        number_of_states: number of choices that has to be generated
+        cum_weights (array[float]): cumulative probabilities of the choices
+        block_size: size of blocks of choices to generate
+    """
 
-        Args:
-            number_of_states (int): number of choices that has to be generated
-            cum_weights (array[float]): cummulative probabilities of the choices
-            block_size (int): size of blocks of choices to generate
-        """
-        if not number_of_states == len(cum_weights):
-            raise Exception('specification of cummulative weights does not match number of states')
+    number_of_states: int
+    cum_weights: np.ndarray
+    block_size: int = 5000
 
-        self.number_of_states = number_of_states
+    def __post_init__(self):
+        if not self.number_of_states == len(self.cum_weights):
+            raise Exception(
+                f'specification of cumulative weights {len(self.cum_weights)} does not match number of states {self.number_of_states}')
         self._idx = 0
-        self._block_size = block_size
-        self.cum_weights = cum_weights
-        self._generate_block()
+        self._block = self._generate_block()
 
-    def _generate_block(self):
-        values = np.random.rand(self._block_size, )
+    def _generate_block(self, size: Optional[int] = None) -> np.ndarray:
+        if size is None:
+            size = self.block_size
+        values = np.random.rand(size, )
         counts, _ = np.histogram(values, [0] + list(self.cum_weights))
-        self._block = np.hstack(tuple([choice_idx * np.ones(c, dtype=int) for choice_idx, c in enumerate(counts)]))
-        np.random.shuffle(self._block)
+        block = np.hstack(tuple(choice_idx * np.ones(c, dtype=int) for choice_idx, c in enumerate(counts)))
+        np.random.shuffle(block)
+        return block
 
-    def generate_choice(self):
-        """ Generate a choice
+    def generate_choice(self) -> int:
+        """ Generate a single choice
 
         Returns:
-            int: integer in the range 0 to the number of states
+            Integer in the range 0 to the number of states
         """
         self._idx = self._idx + 1
-        if self._idx == self._block_size:
```

Comment:
generate_choice doesn't return item 0 for the first generated block. In line 42, self._idx is initialized to 0 and here it is increased before returning the element. while after generating the second block the first item return has self._idx = 0.
I don't know exactly how this is used if this is an issue

Reply:
This is no issue, it just means that we generate one random number that is not used.  We could avoid it, but it would  make the rest of the flow of the nice a bit more involved.